Swap the analyze and detect phases (#412)
buildpacks/docs: content/docs/concepts/components/lifecycle/analyze.md

+++
title="Analyze"
weight=1
summary="Restores files that buildpacks may use to optimize the build and export phases."
+++

{{< param "summary" >}}

> The `analyzer` is responsible for analyzing the cache and the metadata from previously run images (if available) to determine what layers can or cannot be reused.

Reviewer: ```suggestion
Prior to `Platform API 0.7`, the `analyzer` was responsible for analyzing the metadata from the cache and the previously built image (if available) to determine what layers can or cannot be reused.
```

Author: Please see https://github.com/buildpacks/docs/pull/412/commits/f343211c5ae42882b02dfcb04f197b716917e9d4.

This information is used during the `export` phase in order to avoid re-uploading unchanged layers.\

> Starting from `Platform API 0.7`, the `analyze` phase runs before the `detect` phase in order to validate registry access for all images that are used prior to the `detect` and `build` phases, and therefore provides faster failures for end users.\

Reviewer: ```suggestion
Starting from `Platform API 0.7`, the `analyze` phase runs before the `detect` phase in order to validate registry access for all images that are used during the build, and therefore provides faster failures for end users.
```

Reviewer: Ohh is that `\` for adding an extra newline? TIL...

Author: Don't you want to emphasize that in case of a problem, starting from `platform API 0.7`, the user will get an error during the `analyze` phase instead of the `build` phase that comes after the `detect` and `analyze` phases?

Reviewer: I think that's what it's saying now, no?

Author: Please see https://github.com/buildpacks/docs/pull/412#discussion_r706408676 for an idea on how to rewrite this.

Author: I like the addition of "The other responsibilities of the `analyzer` are moved to the `restorer`."

Author: Please see https://github.com/buildpacks/docs/pull/412/commits/74d6a4cee786e2ee6473f37d70c8711759baeacb and let me know what you think.

> Starting from `Platform API 0.7`, the `analyze` phase runs before the `detect` phase in order to validate registry access for all images that are used during the build as early as possible. This validation now happens in the first lifecycle phase, as part of the `analyzer`, before running both the `detector` and the `builder`. In this way it provides faster failures for end users. The other responsibilities of the `analyzer` were moved to the `restorer`.\

Reviewer: ```suggestion
Starting from `Platform API 0.7`, the `analyze` phase runs before the `detect` phase in order to validate registry access for all images that are used during the build as early as possible. In this way it provides faster failures for end users. The other responsibilities of the `analyzer` were moved to the `restorer`.\
```
(analyzer always ran before builder)

Author: I know that the analyzer always ran before the builder, just thought that it's worth emphasizing it, but I don't have strong opinion about this so I'll make this change.
Thanks.

Author: Please see https://github.com/buildpacks/docs/pull/412/commits/3fae08256362ad2fc1293dfba4a5543d1272e2fa.
